# Copyright © 2024 OTIV B.V.
# SPDX-License-Identifier: MIT

diff --git a/src/lib.rs b/src/lib.rs
index 798459e..82c1c43 100644
--- a/src/lib.rs
+++ b/src/lib.rs
@@ -45,4 +45,4 @@ pub const WEBGL2_IDL: &'static [u8] = include_bytes!("../api_webgl/specs/latest/
 /// The contents of the WebGL extension XML files
 /// These are discovered via a build script to avoid having to list each extension by name.
 pub const WEBGL_EXT_XML: &'static [&'static [u8]] =
-    include!(concat!(env!("OUT_DIR"), "/webgl_exts.rs"));
+    include!(env!("WEBGL_EXTS"));
